### DISCUSSION TITLE
Bridge Interface Parameters: Protocol-Mode and VLAN-Filtering

### DISCUSSION THREAD

**Discussion by paulcarre07** [netbox-community/netbox#17798] [Help Wanted!] [answered]
Hello everyone, I'm new to NetBox and working on creating a template for Mikrotik devices. In my template, I need to model a bridge interface, but I'm unsure how to configure the protocol-mode and vlan-filtering settings for a bridge interface. Could anyone provide guidance or examples on how to set this up in NetBox?

**Comment by candlerb** [comment, marked as answer]
That's `protocol-mode=none|stp|rstp|mstp`; `vlan-filtering=yes|no`

To be honest, I would expect that to be hard-wired into the template, which is written for how you want the device to run. If you really need to set them differently on different bridges in different devices, then I guess each bridge would be a bridge interface in Netbox, and you could add a custom field or tag on that interface for those options.

I did experiment with generating some interface configs on Mikrotik a while back, but I am not using them in practice. Here are the templates in case they are useful.

```
{%- set dif = device.interfaces.exclude(type="virtual") %}
{%- for v in device.site.vlans.all() %}
{%- set ift = ((dif.filter(tagged_vlans=v)|list) + (dif.filter(mode="tagged-all")|list)) | unique | sort(attribute="name") %}
{%- set ifu = dif.filter(untagged_vlan=v) | sort(attribute="name") -%}
{%- if ift or ifu %}
/interface bridge vlan add bridge=bridge vlan-ids={{v.vid}}
{%- if ift %} tagged={{ ift|join(",") }}{% endif %}
{%- if ifu %} untagged={{ ifu|join(",") }}{% endif %}
{%- endif %}
{%- endfor %}

--- alternative implementation ---

{%- for v in device.site.vlans.all() %}
{%- set ift = ((v.interfaces_as_tagged.exclude(type="virtual").filter(device=device)|list)
             + (device.interfaces.filter(mode="tagged-all")|list))|sort(attribute="name") %}
{%- set ifu = v.interfaces_as_untagged.exclude(type="virtual").filter(device=device)|sort(attribute="name") %}
{%- if ift or ifu %}
/interface bridge vlan add bridge=bridge vlan-ids={{v.vid}}
{%- if ift %} tagged={{ ift|join(",") }}{% endif %}
{%- if ifu %} untagged={{ ifu|join(",") }}{% endif %}
{%- endif %}
{%- endfor %}
```

`ift` are tagged interfaces, `ifu` are untagged interfaces. Note that in this case I'm just configuring a single fixed bridge, not modelling the bridge(s) themselves within Netbox. If you want multiple bridges then you'd have to iterate over those separately, and then find interfaces linked to them using the "bridge" property.

Jinja2 is pretty ugly for that sort of stuff, and it may be better to do complex templating externally to Netbox.

Since I wrote the above, Netbox has added new top-level model objects in the templates, so you should be able to replace

```
{%- for v in device.site.vlans.all() %}
```

with

```
{%- for v in ipam.VLAN.objects.all() %}
```

(untested) - so that you can find all Netbox VLANs, not just those directly linked to the Device's Site. But you'd need to decide how to deal with `tagged-all` interfaces: all VLANs in the device's site seemed to be a reasonable option.

**Reply by paulcarre07** [reply]
Thank you for your detailed response and for sharing the template examples!

I appreciate your suggestion about hard-wiring the protocol-mode options into the template. My initial thought was to keep the configurations flexible to accommodate various bridge settings across different devices, which is why I considered setting these options dynamically. Your idea of utilizing  custom fields or tags seems like a feasible solution.

**Reply by candlerb** [reply]
Sure. Remember that each device is linked to a config template, so you can write multiple config templates for different purposes and link each device to the appropriate one. You don't need to write a generic "does everything" template.
